File: RemoveInvalidParantheses.java
// Time Complexity : O(n^n)
// Space Complexity : O(n^n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

public class RemoveInvalidParantheses {
    class Solution {
        public List<String> removeInvalidParentheses(String s) {
            if(s == null || s.length() == 0)
                return new ArrayList<>();

            List<String> result = new ArrayList<>();
            Set<String> set = new HashSet<>();
            Queue<String> q = new LinkedList<>();
            q.add(s);
            set.add(s);

            boolean found = false;
            while(!q.isEmpty() && !found){
                int size = q.size();
                for(int i = 0; i < size; i++){
                    String curr = q.poll();
                    if(isValid(curr)){
                        found = true;
                        result.add(curr);
                    }
                    else{
                        if(!found){
                            for(int  j = 0; j < curr.length(); j++){
                                char c = curr.charAt(j);
                                if(Character.isAlphabetic(c))
                                    continue;
                                String child = curr.substring(0, j) + curr.substring(j+1);
                                if(!set.contains(child)){
                                    set.add(child);
                                    q.add(child);
                                }
                            }
                        }
                    }
                }
            }
            return result;
        }

        private boolean isValid(String str){
            int count = 0;
            for(int i = 0; i < str.length(); i++){
                char c = str.charAt(i);
                if(c == '(')
                    count++;
                else if(c == ')'){
                    if(count == 0)
                        return false;
                    count--;
                }
            }
            return count == 0;
        }
    }
}

File: removeInvalidParantheses.py
#Time Complexity : O(2^n) where n is the length of string
#Space Complexity : O(n)
#Did this code successfully run on Leetcode : Yes

class Solution:
    def isValid(self, s):
        count = 0
        for char in s:
            if char == "(":
                count += 1
            elif char == ")":
                if count == 0:
                    return False
                else:
                    count -= 1

        return count == 0

    def removeInvalidParentheses(self, s: str) -> List[str]:
        q = deque([s])
        result = []
        found = False
        seen = set()

        while q and not found:
            for _ in range(len(q)):
                curr = q.popleft()
                if self.isValid(curr): #check if our current parantheses string is valid or not
                    found = True
                    result.append(curr)
                if not found: #if we haven't already found valid parantheses, only then create new children
                    for i in range(len(curr)):
                        if not curr[i].isalpha(): #check if the char is not a letter
                            child = curr[:i] + curr[i+1:] #remove ith index char to make new child
                            if child not in seen:
                                seen.add(child)
                                q.append(child)

        return result
